Add support for pluggable PM policies

ramakrishnapallala opened the pull request:
arch: Kconfig: Add Kconfig options for Low Power States
soc: arm: nrf52: Select the supported Power States
subsys: power: Add support for pluggable PM policies
samples: subsys: power_mgr: Add changes as per PM policy Kconfig

Commit:
subsys: power: Add support for pluggable PM policies

Add support for adding and selecting pluggable PM policies
which can be enabled based on the application needs.

Also added a dummy policy for demonstration purpose which
simply loops over the supported PM states.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

pizi-nordic commented on subsys/power/Kconfig:
```diff
@@ -8,39 +8,16 @@ config PM_CONTROL_OS_LPS
 config PM_CONTROL_OS_LPS_1
 	bool "Platform supports LPS_1"
 	help
 	  Select this option if SoC support LPS_1 state.
```
Side comment: 

For me the naming of states creates some confusion. I really do now know how many states are in fact supported on my platform and how Zephyr names maps to platform capabilities. 

IMHO each platform which supports any kind of power management should define own list of supported states (Example: my platform support LPS_1, LPS_2 and DEEP_SLEEP) and provide some information how they map to the names in datasheet. User should be able to disable some states in Kconfig, however we should not force him to solve multi-selection test about power management.

ramakrishnapallala replied:
@pizi-nordic  I think we can add the individual supported states as Kconfig option at arch level like SYS_POWER_LOW_POWER_STATE_SUPPORTED and SoC implementations can select them based on the SoC specific support.

So, in policy Kconfig we make them depend on the arch level Kconfig, that way we will only have what is supported by a SoC.

pizi-nordic replied:
I think that it is a very good idea.

pizi-nordic commented on subsys/power/policy/Kconfig:
```diff
@@ -0,0 +1,69 @@
+
+choice
+	prompt "PM Policy"
```
I assume that this is in fact "Idle State management".
Could we name things accordingly in order to be prepared for other features, like Frequency/Voltage scaling etc.?

For example:
- Power Management Options:
  - Idle State Management
     - Idle state handling policy
  - CPU Frequency scaling [example]
  - ....

Commit:
arch: Kconfig: Add Kconfig options for Low Power States

Add Kconfig options for Low Power and Deep Sleep States which
need to be selected by the SoC's based on it's supported
Low Power States.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
soc: arm: nrf52: Select the supported Power States

Select the Low Power States supported by nrf52 SoC series.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
samples: subsys: power_mgr: Add changes as per PM policy Kconfig

Enable the CPU Residency based PM policy Kconfig option for
the sample application.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@pizi-nordic I have updated the PR with proposed changes. Please review the changes.
